### ISSUE TITLE
Make  `makeOperationId()` configurable

### ISSUE THREAD

**Issue by jonaskello** [seriousme/fastify-openapi-glue#423]
When an `operationId` is missing one is constructed automatically by `makeOperationId()` [here](https://github.com/seriousme/fastify-openapi-glue/blob/master/lib/ParserBase.js#L8).

In my case I am using typescript with a mapped type that creates a type for the service object. When an `operationId` is missing I map this to a simple function name. Because mapped types is limited I cannot mimic the advanced logic in `makeOperationId()` so I map to a simpler name. In this case I also need [fastify-openapi-glue](https://github.com/seriousme/fastify-openapi-glue) to use the same name that I create in my mapped type.

One way to achieve this would be to make it possible to supply your own `makeOperationId()` for fastify-openapi-glue. Basically adding an override for that function to the options like this:

```js
const options = {
  specification: `${currentDir}/petstore-openapi.v3.json`,
  service: new Service(), // The service object is using my mapped type and so have special names in the type for missing operationId
  makeOperationId:  (operation, path) => { /* any logic needed to create the missing operationId */ }
};
```

**Comment by seriousme** [comment]
Hi,

thanks for asking. 

The [operationResolver](https://github.com/seriousme/fastify-openapi-glue#options) might be what you need.
Using this option you can basically map any operationId generated by fastify-openapi-glue to any name you like.
But even more easy would be explictly add your own operationID to the openApi spec ;-)

Hope this helps.

Kind regards,
Hans

**Comment by jonaskello** [comment]
Yes, I saw the `operationResolver` and it is close to what I want. However the problem is that once the `operationResolver` is called I cannot distinguish the cases when `operationId`  was actually specificed in the open api spec, or if it was "invented" by fastify-openapi-glue.

It is perhaps easier to explain with an example :-). So let's say we have this spec.yaml:

```yaml
paths:
  /projects/{projectNumber}:
    get:
      operationId: getByProjectNumber
      summary: Get a project by number.
  /customers/{customerNumber}:
    get:
      summary: Get a customer by number.
```` 

So in the case of `/projects/{projectNumber}` there is an `operationId` specified so I the type I am constructing will have that name in it's interface. But in the case of `/customers/{customerNumber}` there is no `operationId` so I will just map this to the name `/customers/{customerNumber}#get`. So when the type is built it looks like this:

```ts
interface {
   getByProjectNumber: Handler;
   "/customers/{customerNumber}#get": Handler;
}
```

And then the service is implemented like this type. Now when `operationResolver()` is called with `getByProjectNumber` I could call this member of the service object and all works. However then `operationResolver()` will be called by `getCustomersByCustomerNumber `. In this case the operationId was "invented" by fastify-openapi-glue but I have no way of telling this is the case. So I cannot map to the  `"/customers/{customerNumber}#get"` member of the service object.

An alternative to an option for `makeOperationId` would be to instead extend the parameters of `operationResolver` so it gets the same parameters as `makeOperationId`. Then it would be possible to map to the correct name. So something like this:

```
const options = {
  specification: `${currentDir}/petstore-openapi.v3.json`,
  service: new Service(), 
  operationResolver: (operation, path) => handler.
};
```

And yes, of course the everything would have been simpler if the open-api spec just required every operation to have an operationId :-).

**Comment by seriousme** [comment]
Hi,

Thanks for elaborating a bit on the challenge you have.
I am still a bit confused as to why you seem to be able to fetch the existing operationId's from the spec and not be able to mimic the generation of missing operationId's (like fastify-openapi-glue does) in your own code.
 
Removing the operationID from the operationResolver would be a breaking change, however `method` and `path` can be added without exposing internal details of fastify-openapi-glue.

Could you check if the [extend_operation_resolver](https://github.com/seriousme/fastify-openapi-glue/tree/extend_operation_resolver) branch does what you need ? The README.md there explains the new parameters to the operationResolver option ;-)

Also you got me quite curious on what your project entails ;-) : is it intended to do something similar to the fastify-openapi-glue `generator` but then for Typescript ?

Kind regards,
Hans

**Comment by jonaskello** [comment]
Yes, I think what you have in the extend_operation_resolver branch would work well for me :-). Then I can find my handlers based on the original path and method from the path.

My project is about doing schema-first development with typescript. So I want to generate an interface for the service object from the oas yaml document. The most straightforward way to do this would of course be to just write a code-gen tool that does that. However since there is already a code-gen tool available called [openapi-typescript](https://www.npmjs.com/package/openapi-typescript) I want to use that instead of writing and maintaining my own tool.

The issue though is that openapi-typescript generates a type that is a 1:1 representation of the schema (rather than interfaces for service objects). For my example schema above it will generate something like this:

```ts
export type paths = {
  readonly "/projects/{projectNumber}": {
    readonly get: {
      readonly parameters: {
        readonly path: {
          readonly projectNumber: string;
        };
      };
      readonly responses: {
        readonly 200: {
          readonly content: {
            readonly "text/plain": string;
          };
        };
        readonly 404: {
          readonly content: {
            readonly "text/plain": string;
          };
        };
      };
    };
  };
  readonly "/customers/{customerNumber}": {
    readonly get: {
      readonly parameters: {
        readonly path: {
          readonly projectNumber: string;
        };
      };
      readonly responses: {
        readonly 200: {
          readonly content: {
            readonly "text/plain": string;
          };
        };
        readonly 404: {
          readonly content: {
            readonly "text/plain": string;
          };
        };
      };
    };
  };
};
```

So to get the interface for the service object I use a typescript feature called [mapped types](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html) that can transform one type into another type. So I do something like this:

```ts
export type MyServiceObj = ServiceObjMapper<paths>; // Where paths is the type generated above
```

The mapped type `MyServiceObj` becomes something like this:

```ts
export type Handler = (request, reply) => unknown;

export type ThisIsWhatTheMappedTypeLooksLike = {
    "/projects/{projectNumber}#get": Handler
    "/customers/{customerNumber}#get": Handler
}
```

So the magic happens in the `ServiceObjMapper` which do all the mapping. It is quite involved but here is the part where it flattens the `path` hiearachy and creates the operation keys:

```ts
export type FlattenSpecPaths<T> = FlattenStepTwo<FlattenStepOne<T>>;
type FlattenStepOne<T> = {
  readonly [K in keyof T as `${string & K}#${keyof T[K] & string}`]: T[K];
};
type FlattenStepTwo<T> = {
  readonly [a in keyof T]: Value2<{
    readonly [M in keyof T[a] as M extends Tail2<a> ? M : never]: T[a][M] extends SpecOperationShape ? T[a][M] : never;
  }>;
};
type Value2<T> = T[keyof T];
type Tail2<S> = S extends `${string}#${infer T}` ? Tail2<T> : S;
```

So this is using another typescript feature called [template-literal-types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) where you can construct new names in the mapped type. This is why I cannot have advanded constructed operationId, becuase it is not constructed by javascript but instead meta-programming of types using typescript features.

**Comment by jonaskello** [comment]
Thinking a bit about performance, perhaps the lookup that `operationResolver` needs to do for each request will add some overhead. Since I cannot be sure if the original spec had an operationId or not, I would have to do a lookup on the service object for the operationId first, then if that fail, construct the "invented" operationId and do a lookup for that.

If instead `makeOperationId` was customizable then I guess the operationsIds would be determined only once and the lookup for each request could be skipped? Adding `makeOperationId` function to options should also be a backwards compatible change.

That being said, I don't think the lookup overhead will have a big impact so either way would work.

**Comment by jonaskello** [comment]
Just realized that I can know if the original spec had an `operationId` since then `operationResolver` is called with an  `operationId` of  `undefined`? In that case I can skip one of the lookups.

**Comment by seriousme** [comment]
Thanks for the elaborate explanation ! Maybe we can add Typescript as an option to the generator as well :-)
 
The handler lookup is done at route generation time, not at route invocation time , so the performance hit should be neglible anyway :-)

I just released 4.1.0 on NPM containing the new parameters.

Enjoy,

Hans

**Comment by jonaskello** [comment]
Thanks for the quick handling of this :-).

I think the typescript definitions needs to be updated to follow the new signature of operationResolver:

https://github.com/seriousme/fastify-openapi-glue/blob/master/index.d.ts#L5

Regarding typscript generation it might be interesting. At least I would probably be able to contribute a mapped type that can create a types for the handlers that fastify-openapi-glue requires from what openapi-typescript generates. I also noticed that  openapi-typescript has added support for operationId so perhaps I will end up requiring that operationId is always provided instead of constructing it as it would be less complex (as you originally suggested :-)).

**Comment by seriousme** [comment]
Thanks for spotting the missing TS signature update, fix is in the works as we speak ;-)

I was thinking to let a generator for Typescript create a fastify plugin that does not require fastify-openapi-glue during runtime anymore.
That way we can make stuff way more simple by doing something like: https://www.fastify.io/docs/latest/Reference/TypeScript/#plugins

However I'm no Typescript guru, so what do you think ?

Kind regards,
Hans

**Comment by seriousme** [comment]
FYI: Fix is in (#426 ) and publish on NPM as 4.1.1
